langchain/agents/agent_toolkits/gmail/toolkit.py

```diff
@@ -0,0 +1,363 @@
+from __future__ import annotations
+from typing import List, Any
+import os
+import base64
+import re
+import json
+from bs4 import BeautifulSoup
+
+import email
+from email.message import EmailMessage
+from email.mime.text import MIMEText
+from email.mime.multipart import MIMEMultipart
+
+from googleapiclient.discovery import build
+from google.oauth2.credentials import Credentials
+from google.auth.transport.requests import Request
+from google.oauth2.credentials import Credentials
+from google_auth_oauthlib.flow import InstalledAppFlow
+
+from langchain.agents.tools import Tool
+from langchain.agents.agent_toolkits.base import BaseToolkit
+from langchain.tools import BaseTool
+from langchain.llms.base import BaseLLM
+from langchain.agents.agent_toolkits.gmail.prompt import (
+    QUERY_PROMPT,
+    DRAFT_PROMPT,
+    SEND_PROMPT,
+)
+from langchain.prompts import PromptTemplate
+
+from datetime import datetime
+
+
+SCOPES = ["https://mail.google.com/"]
+
+
+class GmailToolkit(BaseToolkit):
```

Comment:
What we've done in the past is refactor all of the client methods outisde of the toolkit in a separate wrapper (see zapier, sql, etc). Then we've initialized the toolkit with that wrapper. It makes the things a bit more modular. Additionally, it looks like some of these methods can throw. One of the benefits of using an agent for these things is retrying and correcting for errors, so could be worth stringifying the error messages for the tools and allowing the agent to interpret them